### ISSUE TITLE
scalafmt as gradle plugin

### ISSUE THREAD

**Issue by Nutomic** [scalameta/scalafmt#139]
This way, we can format our code on each compile automatically, similar to the SBT plugin.

https://docs.gradle.org/current/userguide/plugins.html
https://docs.gradle.org/current/userguide/custom_plugins.html

**Comment by olafurpg** [comment]
Is there a scalariform (or similar) plugin that can be reused except with scalafmt? Is it possible to automatically upgrade the plugin as part of a [publish script](https://github.com/olafurpg/scalafmt/blob/master/bin/publish.sh)?

I don't use gradle personally so I would think it is better for someone who uses it to be involved.

**Comment by Nutomic** [comment]
I think the plugins are distributed through Maven Central.

I don't know scalariform (and couldn't find it for gradle in a quick search), but from my links above, it seems to be pretty easy to add a new gradle task that just calls into a jar. That way we could run `gradle scalafmt assemble` to format and build.

**Comment by olafurpg** [comment]
Sounds great, I'm sure it's not a lot of work for someone with gradle experience. It would require more effort for me since I have never used gradle!

I'm happy to assist anyone who wants to contribute such a plugin. We can make it part of the scalafmt build so the plugin is released alongside scalafmt.

**Comment by olafurpg** [comment]
It might be possible to reuse some parts of this plugin: https://github.com/hierynomus/scalariform-gradle-plugin

**Comment by alenkacz** [comment]
working on it :) give me few days... https://github.com/alenkacz/gradle-scalafmt

**Comment by alenkacz** [comment]
btw. the plugin is done from my point of view. It does not support everything that sbt does but the most important stuff are there. I am already using it in one of my projects and it works well...

**Comment by olafurpg** [comment]
Cool @alenkacz! Thank you for sharing. Would you mind if I include your plugin under ["Related projects"](https://olafurpg.github.io/scalafmt/#Relatedprojects) in the docs?

**Comment by olafurpg** [comment]
BTW, out of curiosity, did you consider using `scalafmt-cli`? The sbt plugin is a thin wrapper around the command line interface. For example, here's how the `scalafmtTest`/`scalafmtDiff` tasks are implemented: https://github.com/olafurpg/scalafmt/blob/master/scalafmtSbt/src/main/scala/org/scalafmt/sbt/ScalafmtPlugin.scala#L30-L32

**Comment by alenkacz** [comment]
Yes, feel free to put it there. Thanks!

I did consider it. I chose to do it as I did for 2 reasons:
- I can achieve more "gradle-like" feel and behavior this way
- it is just ~50 lines of code to do what I am doing there which is maintainable :)

I can switch to CLI in the future, if there is a strong reason to do that (e.g. I would have to change my implementation too often to keep up with changes in your library)
